### PR TITLE
[Dynamic Dashboard] In-app feedback card

### LINKED ISSUE #12600: Add feedback banner for rating the app
- Check with local storage if the banner should be displayed using `AppGeneralAction.loadFeedbackVisibility` (following the logic [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsAndTopPerformersPeriodViewModel.swift))
- Display the banner if needed in the 2nd place on the dashboard.
- Rebuild the banner or embed `InAppFeedbackCardViewController` depending on complexity.

Internal - p1714408022255559/1712135440.001229-slack-C03L1NF1EA3

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #12600

## Description

Show the in-app feedback card on the SwiftUI-based dashboard screen. 

Changes
- Move logic for checking when to show in app feedback card into `DashboardViewModel`.
- Write InAppFeedback card view using SwiftUI
- Show the card and survey from `DashboardView`

## Testing instructions

**Survey flow**
- Do a fresh install of the app on your device
- Login into the app
- Change the device date to any date after 90 days
- Open the My store tab. Switch tabs to refresh the screen if needed.
- Validate that you see the In-App feedback card.
- Tap on "Could be better" and validate that the following event is tracked
```
Tracked app_feedback_prompt, properties: [AnyHashable("store_id"): "123", AnyHashable("blog_id"): 123, AnyHashable("is_wpcom_store"): false, AnyHashable("plan"): "jetpack_security_daily", AnyHashable("action"): "didnt_like", AnyHashable("was_ecommerce_trial"): false, AnyHashable("site_url"): "https://example.com"]
```
- A survey screen should be presented and the following event should be tracked
```
Tracked survey_screen, properties: [AnyHashable("blog_id"): 123, AnyHashable("was_ecommerce_trial"): false, AnyHashable("action"): "opened", AnyHashable("store_id"): "123", AnyHashable("is_wpcom_store"): false, AnyHashable("context"): "general", AnyHashable("plan"): "jetpack_security_daily", AnyHashable("site_url"): "https://example.com"]
```
- The in-app feedback card should be removed from the dashboard.

**Rating flow**
- Do a fresh install of the app on your device
- Login into the app
- Stop the app from Xcode
- Change `shouldBeVisible` into `true` in this line https://github.com/woocommerce/woocommerce-ios/pull/12636/files#diff-5e4b4001de415046921cf84e4ce2b6c16646ae1bb2067aa6d9ec1fbfa2a855e5R413 to always show the in-app feedback card. Code change is required as changing Date disrupts the App store rating part due to certificate expired error. 
- Run the app from Xcode with the altered code.
- Open the My store tab. Switch tabs to refresh the screen if needed.
- Validate that you see the In-App feedback card.
- Tap on "I like it" and validate that the following event is tracked
```
Tracked app_feedback_prompt, properties: [AnyHashable("store_id"): "123", AnyHashable("blog_id"): 123, AnyHashable("is_wpcom_store"): false, AnyHashable("plan"): "jetpack_security_daily", AnyHashable("action"): "liked", AnyHashable("was_ecommerce_trial"): false, AnyHashable("site_url"): "https://example.com"]
```
- App rating popup should be presented.

## Screenshots

| Dark | Light |
|--------|--------|
| IMG_80BC3E7E98C5-1 | IMG_BE43BE952EC3-1 | 

Survey flow
https://github.com/woocommerce/woocommerce-ios/assets/524475/0f0cd4a4-1b22-4347-b473-d17d2cc29ded

Rating flow

https://github.com/woocommerce/woocommerce-ios/assets/524475/992305ab-4aa9-4be3-b2d0-ec8836e2f237

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by hafizrahman** [comment]
I can't make the rating view appear, so could use your direction to check what I'm missing.

Here's my steps:

1. Delete app on real device,
2. Build app on real device,
3. Login,
4. First thing I see is the Feature Announcement,
5. Go to Date & Setting, change device date to December 2024,
6. Go back to app, close Feature Announcement, refresh My Store,
7.  In-App feedback card is shown,
8. Tap "Could be better" (trackings are shown correctly)
9. Survey screen shown,
10. Close survey screen, In-app feedback card is hidden,
11. Go back to Xcode, update `.sendIsInAppFeedbackCardVisibleValueAndTrackIfNeeded(true)`
12. Build app again,
13. My Store now shows in-app feedback card again,
14. Tap "Could be better",
15. Survey screen is shown again, instead of the rating view.

File: WooCommerce/Classes/ViewRelated/InAppFeedback/InAppFeedbackCardViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    func didTapCouldBeBetter() {
+        analytics.track(event: .appFeedbackPrompt(action: .didntLike))
+        onFeedbackGiven?(.didntLike)
+    }
+
+    func didTapILikeIt() {
+        analytics.track(event: .appFeedbackPrompt(action: .liked))
+        if let windowScene = UIApplication.shared.connectedScenes.first(where: { $0.activationState == .foregroundActive}) as? UIWindowScene {
+            storeReviewControllerType.requestReview(in: windowScene)
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/InAppFeedback/InAppFeedbackCardViewModel.swift:41]: app rating seems to only be called on "I Like It" and never on "Could be better" and I'm not sure what's the correct behavior, based on the test instruction this seems to need to be called by "Could be better" too.

**Comment by itsmeichigo** [comment]
Just chiming in: I believe we should remove `InAppFeedbackCardViewController` and the related xib file too as they are no longer used.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/InAppFeedback/InAppFeedbackCardViewModel.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/InAppFeedback/InAppFeedbackCardViewModel.swift:41]: Thanks, Hafiz! I updated the test instructions for the app ratings flow.

**Comment by selanthiraiyan** [comment]
Thanks for the review, @hafizrahman!

Before testing the app rating flow, the app needs to be deleted and installed again. I'm sorry; I updated the test instructions now. 

Kindly take another look at the PR when you can.

**Comment by selanthiraiyan** [comment]
Thanks @itsmeichigo! Good point. 

> Just chiming in: I believe we should remove InAppFeedbackCardViewController and the related xib file too as they are no longer used.

Done in https://github.com/woocommerce/woocommerce-ios/pull/12636/commits/a2fdb4fb64b8c3d1a5717a27ff49b64803662c52

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 2]
```diff
@@ -381,6 +401,57 @@ private extension DashboardViewModel {
...
+    /// Calculates and updates the value of `isInAppFeedbackCardVisible`.
+    func refreshIsInAppFeedbackCardVisibleValue() {
+        let action = AppSettingsAction.loadFeedbackVisibility(type: .general) { [weak self] result in
+            guard let self = self else {
+                return
+            }
+
+            switch result {
+            case .success(let shouldBeVisible):
+                self.sendIsInAppFeedbackCardVisibleValueAndTrackIfNeeded(true)
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: I think this part is accidentally changed to true.

**Review by hafizrahman — APPROVED** [review]
This now works as expected in my test and the code also makes sense but I left a code comment. Pre-approving for now, please feel free to tackle and merge after that.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Great catch.  Fixed in https://github.com/woocommerce/woocommerce-ios/pull/12636/commits/49a320cb59986a0a8abf4e1787c2791d53c289b1

**Comment by selanthiraiyan** [comment]
Thanks for the review, @hafizrahman! I sneaked in a commit which fixes something that I missed while resolving merge conflicts. https://github.com/woocommerce/woocommerce-ios/pull/12636/commits/5814f400f940de2d00a1a312640e23c15b55574c Enabling auto-merge now. 🙇

**Pull request merged**